File: src/main/java/hzt/controller/sub_pane/AppearanceController.java
package hzt.controller.sub_pane;

import hzt.controller.FXMLController;
import hzt.controller.SceneManager;
import hzt.controller.scenes.MainSceneController;
import hzt.controller.scenes.SceneController;
import hzt.model.Resource;
import hzt.service.BackgroundService;
import hzt.service.IBackgroundService;
import hzt.service.IThemeService;
import hzt.service.ThemeService;
import javafx.beans.value.ObservableValue;
import javafx.collections.ObservableList;
import javafx.fxml.FXML;
import javafx.geometry.Insets;
import javafx.scene.control.ColorPicker;
import javafx.scene.control.ComboBox;
import javafx.scene.control.ComboBoxBase;
import javafx.scene.control.ToggleButton;
import javafx.scene.image.Image;
import javafx.scene.input.KeyCode;
import javafx.scene.input.KeyEvent;
import javafx.scene.layout.AnchorPane;
import javafx.scene.layout.Background;
import javafx.scene.layout.BackgroundFill;
import javafx.scene.layout.BackgroundImage;
import javafx.scene.layout.BackgroundPosition;
import javafx.scene.layout.BackgroundRepeat;
import javafx.scene.layout.BackgroundSize;
import javafx.scene.layout.CornerRadii;
import javafx.stage.Stage;

import java.io.IOException;
import java.util.Collection;
import java.util.Optional;

import static hzt.model.AppConstants.INIT_BG_COLOR;
import static java.lang.Boolean.TRUE;

public class AppearanceController extends FXMLController {

    @FXML
    private ToggleButton opacityStageButton;
    @FXML
    private ComboBox<Resource> backgroundCombobox;
    @FXML
    private ComboBox<Resource> themeCombobox;
    @FXML
    private ToggleButton fullScreenButton;
    @FXML
    private ColorPicker backgroundColorPicker;

    private final IThemeService themeService = new ThemeService();
    private final IBackgroundService backgroundService = new BackgroundService();
    private final MainSceneController mainSceneController;

    public AppearanceController(MainSceneController mainSceneController) throws IOException {
        super("appearancePane.fxml");
        this.mainSceneController = mainSceneController;
        configureComboBoxes();
        configureColorPickers();
        configureStageControlButtons();
    }

    private void configureColorPickers() {
        backgroundColorPicker.setValue(INIT_BG_COLOR);
        mainSceneController.getBackgroundColorPicker().valueProperty()
                .bindBidirectional(backgroundColorPicker.valueProperty());
    }

    public final void configureStageControlButtons() {
        Stage stage = mainSceneController.getSceneManager().getStage();
        fullScreenButton.setOnAction(e -> stage.setFullScreen(!stage.isFullScreen()));
        stage.fullScreenProperty().addListener((o, c, n) -> fullScreenButton.setSelected(n));
        stage.addEventFilter(KeyEvent.KEY_PRESSED, key -> setFullScreenWhenF11Pressed(stage, key));
        bindStageOpacityToOpacityButton(stage);
    }

    private static void setFullScreenWhenF11Pressed(Stage stage, KeyEvent keyEvent) {
        if (keyEvent.getCode() == KeyCode.F11) {
            stage.setFullScreen(!stage.isFullScreen());
        }
    }

    private void bindStageOpacityToOpacityButton(Stage stage) {
        opacityStageButton.selectedProperty().addListener((o, c, n) -> stage.setOpacity(TRUE.equals(n) ? .8 : 1));
    }

    private void configureComboBoxes() {
        themeCombobox.getItems().addAll(themeService.getThemes());
        themeService.currentThemeProperty().bind(themeCombobox.valueProperty());
        themeService.styleSheetProperty().addListener(this::changeStyle);
        themeCombobox.setValue(IThemeService.DEFAULT_THEME);

        backgroundCombobox.getItems().addAll(backgroundService.getResources());
        backgroundCombobox.setValue(BackgroundService.NO_PICTURE);
    }


    public void changeStyle(ObservableValue<? extends String> o, String c, String newVal) {
        SceneManager sceneManager = mainSceneController.getSceneManager();
        Collection<SceneController> sceneControllers = sceneManager.getSceneControllerMap().values();
        for (SceneController sceneController : sceneControllers) {
            ObservableList<String> styleSheets = sceneController.getScene().getStylesheets();
            styleSheets.removeIf(filter -> !styleSheets.isEmpty());
            if (newVal != null) {
                styleSheets.add(newVal);
            }
        }
    }

    @FXML
    private void backgroundComboBoxAction() {
        Optional.of(backgroundCombobox)
                .map(ComboBoxBase::getValue)
                .map(Resource::getInputStream)
                .map(Image::new)
                .ifPresentOrElse(this::setBackgroundImage, this::backgroundColorPickerAction);
    }

    @FXML
    private void backgroundColorPickerAction() {
        mainSceneController.getAnimationPane().setBackground(new Background(
                new BackgroundFill(backgroundColorPicker.getValue(), CornerRadii.EMPTY, Insets.EMPTY)));
        backgroundCombobox.setValue(BackgroundService.NO_PICTURE);
    }

    @Override
    protected FXMLController getController() {
        return this;
    }

    private void setBackgroundImage(Image image) {
        AnchorPane animationPane = mainSceneController.getAnimationPane();
        animationPane.setBackground(new Background(new BackgroundImage(image,
                BackgroundRepeat.NO_REPEAT,
                BackgroundRepeat.NO_REPEAT,
                BackgroundPosition.CENTER,
                new BackgroundSize(animationPane.getWidth(), animationPane.getHeight(),
                        false, false, false, true))));
    }
}

File: src/main/java/hzt/controller/sub_pane/AppearanceController.kt
package hzt.controller.sub_pane

import hzt.controller.FXMLController
import hzt.controller.scenes.MainSceneController
import hzt.controller.scenes.SceneController
import hzt.model.Resource
import hzt.service.BackgroundService
import hzt.service.IBackgroundService
import hzt.service.IThemeService
import hzt.service.ThemeService
import hzt.utils.inverseFullScreen
import hzt.utils.onNewValue
import javafx.event.EventHandler
import javafx.fxml.FXML
import javafx.geometry.Insets
import javafx.scene.control.ColorPicker
import javafx.scene.control.ComboBox
import javafx.scene.control.ToggleButton
import javafx.scene.image.Image
import javafx.scene.input.KeyCode
import javafx.scene.input.KeyEvent
import javafx.scene.layout.*
import javafx.scene.paint.Color
import javafx.stage.Stage
import java.util.*

class AppearanceController(private val mainSceneController: MainSceneController) :
    FXMLController("appearancePane.fxml") {

    @FXML
    private lateinit var opacityStageButton: ToggleButton
    @FXML
    private lateinit var backgroundCombobox: ComboBox<Resource>
    @FXML
    private lateinit var themeCombobox: ComboBox<Resource>
    @FXML
    private lateinit var fullScreenButton: ToggleButton
    @FXML
    private lateinit var backgroundColorPicker: ColorPicker

    private val themeService: IThemeService = ThemeService()
    private val backgroundService: IBackgroundService = BackgroundService()

    init {
        configureComboBoxes()
        configureColorPickers()
        configureStageControlButtons()
    }

    private fun configureColorPickers() {
        backgroundColorPicker.value = INIT_BG_COLOR
        mainSceneController.backgroundColorPicker.valueProperty()
            .bindBidirectional(backgroundColorPicker.valueProperty())
        backgroundColorPicker.valueProperty().onNewValue(::backgroundColorPickerAction)
    }

    private fun configureStageControlButtons() {
        val stage = mainSceneController.sceneManager.stage
        fullScreenButton.configure(stage)
        stage.bindOpacityToOpacityButton(opacityStageButton)
    }

    private fun ToggleButton.configure(stage: Stage) {
        mainSceneController.fullScreenButton.selectedProperty().bindBidirectional(selectedProperty())
        onAction = EventHandler { stage.inverseFullScreen() }
        stage.fullScreenProperty().onNewValue { isSelected = it }
        stage.addEventFilter(KeyEvent.KEY_PRESSED) {
            if (it.code == KeyCode.F11) {
                stage.inverseFullScreen()
            }
        }
    }

    private fun Stage.bindOpacityToOpacityButton(opacityStageButton: ToggleButton) =
        opacityStageButton.selectedProperty().onNewValue { opacity = if (it) .8 else 1.0 }

    private fun configureComboBoxes() {
        themeCombobox.items.addAll(themeService.themes)
        themeService.currentThemeProperty().bind(themeCombobox.valueProperty())
        themeService.styleSheetProperty().onNewValue(::changeStyle)
        themeCombobox.value = IThemeService.DEFAULT_THEME
        backgroundCombobox.items.addAll(backgroundService.resources)
        backgroundCombobox.value = BackgroundService.NO_PICTURE
    }

    private fun changeStyle(newVal: String?) {
        val sceneManager = mainSceneController.sceneManager
        val sceneControllers: Collection<SceneController> = sceneManager.sceneControllerMap.values
        for (sceneController in sceneControllers) {
            val styleSheets = sceneController.scene.stylesheets
            styleSheets.removeIf { styleSheets.isNotEmpty() }
            Optional.ofNullable(newVal).ifPresent(styleSheets::add)
        }
    }

    @FXML
    private fun backgroundComboBoxAction() = Optional.of(backgroundCombobox)
        .map { it.value }
        .map(Resource::getInputStream)
        .map(::Image)
        .ifPresentOrElse(::setBackgroundImage) { backgroundColorPickerAction(backgroundColorPicker.value) }

    private fun backgroundColorPickerAction(newColor: Color) {
        backgroundCombobox.value = BackgroundService.NO_PICTURE
        mainSceneController.animationPane.background = Background(
            BackgroundFill(newColor, CornerRadii.EMPTY, Insets.EMPTY)
        )
    }

    override fun getController(): FXMLController = this

    private fun setBackgroundImage(image: Image) {
        val animationPane = mainSceneController.animationPane
        animationPane.background = Background(
            BackgroundImage(
                image, BackgroundRepeat.NO_REPEAT, BackgroundRepeat.NO_REPEAT, BackgroundPosition.CENTER,
                BackgroundSize(animationPane.width, animationPane.height, false, false, false, true)
            )
        )
    }

    companion object {
        val INIT_BG_COLOR = Color.NAVY
    }
}
